runtime/memprof.c

```diff
@@ -42,45 +550,612 @@ struct memprof_thread_s {
+       * to fake deallocating the block, so that alloc/dealloc
+       * callback counts correspond. */
+      if ((es->t[i].callbacks & CB_MASK(CB_ALLOC)) &&
+          ((es->t[i].callbacks & CB_MASK(CB_DEALLOC)) == 0)) {
+        es->t[i].offset = false;
+        es->t[i].block = Val_unit;
+        es->t[i].deallocated = 1;
+        if (i < es->next) es->next = i;
+      } else {
+        dead = true;
+      }
+    }
+
+    if (dead) {
+      /* entry is evicted, but what if a thread is running it?  This
+       * can happen specifically if an allocation callback stops a
+       * profile and starts a new one: all 'offset' entries are evicted.
+       */
+      memprof_thread_t runner = es->t[i].running;
+      if (runner != NULL) {
+        CAMLassert(runner->running_table == es);
+        CAMLassert(runner->running_index == i);
+        runner->running_table = NULL;
+      }
+    } else { /* preserve this entry */
+      if (i != j) {
+        es->t[j] = es->t[i];
+        /* if a callback is currently being run on this entry,
+         * make sure its index is updated */
+        memprof_thread_t runner = es->t[i].running;
+        if (runner != NULL) {
+          CAMLassert(runner->running_table == es);
+          CAMLassert(runner->running_index == i);
+          runner->running_index = j;
+        }
+      }
+      ++ j;
+    }
+    ++ i;
+    if (es->young == i) es->young = j;
+    if (es->next == i) es->next = j;
+  }
+  es->evict = es->live = j;
+  CAMLassert(es->next <= es->live);
+  CAMLassert(es->young <= es->live);
+
+  ensure_entries(es, 0);
+}
+
+/* Transfer all entries from one entries table to another. */
+
+static void entries_transfer(entries_t from, entries_t to)
+{
+  size_t offset = to->live;
+
+  if (from->live == 0) {
+    return;
+  }
+
+  CAMLassert(to->config == from->config);
+
+  ensure_entries(to, from->live);
+  to->live += from->live;
+
+  for (size_t j = 0; j < from->live; ++j) {
+    to->t[offset + j] = from->t[j];
+    memprof_thread_t thread = from->t[j].running;
+    if (thread) { /* unusual */
+      CAMLassert(thread->running_table == from);
+      CAMLassert(thread->running_index == j);
+      thread->running_table = to;
+      thread->running_index = offset+j;
+    }
+  }
+
+  if (to->young == offset) {
+    to->young = offset + from->young;
+  }
+  if (to->evict == offset) {
+    to->evict = offset + from->evict;
+  }
+  if (to->next == offset) {
+    to->next = offset + from->next;
+  }
+  /* Reset `from` to empty, and allow it to shrink */
+  ensure_entries(from, -from->live);
+  from->young = from->evict = from->next = from->live = 0;
+}
+
+/* If es->config points to a DISCARDED configuration, update
+ * es->config to CONFIG_NONE. Return es->config. */
+
+static value validated_config(entries_t es)
+{
+  value config = es->config;
+  if ((config != CONFIG_NONE) &&
+      (Status(config) == CONFIG_STATUS_DISCARDED)) {
+    es->config = config = CONFIG_NONE;
+  }
+  return config;
```

Comment:
```suggestion
  if ((es->config != CONFIG_NONE) &&
      (Status(es->config) == CONFIG_STATUS_DISCARDED)) {
    es->config = CONFIG_NONE;
  }
  return es->config;
```

This improves readability, and two more in-cache memory accesses should really be negligible.

Reply:
Also, in the case an entry table is found discarded, perhaps we should free some memory by removing all the entries which are not running a callback. If memprof is started and then discarded, the domain entries tables can occupy a significant amount of memory which is never freed otherwise.

This could be done by a special case in `entries_evict` which would check that the entry table is discarded, and if so removes all entries not running a callback (entries running a callback occupy a negligible amount of memory).